MiniApp/Classes/MiniAppClient.swift

```diff
@@ -1,50 +1,66 @@
-/**
- * Protocol for the client that talks to 'GET /app:/version:/manifest' endpoint.
- */
-protocol MiniAppClientProtocol {
+struct ResponseData {
+    let data: Data
+    let httpResponse: HTTPURLResponse
+    init(_ data: Data, _ httpResponse: HTTPURLResponse) {
+        self.data = data
+        self.httpResponse = httpResponse
+    }
+}
 
-    /// Fetches the manifest from the MiniApp backend.
-    /// - Parameters:
-    ///   - appId: String - AppID of the MiniApp.
-    ///   - versionId: String - VersionID of the MiniApp.
-    /// - Returns:Response of from the endpoint.
-    func fetchManifest(with appId: String, and versionId: String) -> ManifestResponse?
+struct ErrorData: Decodable, Equatable {
+    let code: Int
+    let message: String
 }
 
-struct MiniAppClient: MiniAppClientProtocol {
-    var serviceCommunicator: ServiceCommunicatorProtocol = ServiceCommunicator()
+fileprivate extension NSError {
+    class func serverError(code: Int, message: String) -> NSError {
+        return NSError(domain: "Server", code: code, userInfo: [NSLocalizedDescriptionKey: message])
+    }
+}
 
-    func fetchManifest(with appId: String, and versionId: String) -> ManifestResponse? {
-        guard let manifestUrl = formRequestUrl(with: appId, and: versionId) else {
-            #if DEBUG
-                print("MiniAppSDK: Failed to create manifest URL.")
-            #endif
-            return nil
-        }
+protocol MiniAppClientProtocol {
+    func startDataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void)
+}
 
-        let (responseBody, metadata) = serviceCommunicator.requestFromServer(withUrl: manifestUrl,
-                                                                             withHttpMethod: .get,
-                                                                             withSemaphoreWait: true)
+extension URLSession: MiniAppClientProtocol {
+    func startDataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) {
```

Comment:
Can you use a `Result` type for the completion handler and handle the optionals inside the function?

Reply:
Yes we can. I was thinking the same to have Result type, but as of now dataTask methods of URLSession doesn't support any result types and by doing above changes I might end up doing most of the error handling in the extension itself (which is good though). But I thought I should follow similar standards like Remote config.